org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/RemoteEndpoint.java

```diff
@@ -164,35 +182,50 @@ public void consume(Message message) {
 		} else if (message instanceof ResponseMessage) {
 			ResponseMessage responseMessage = (ResponseMessage) message;
 			handleResponse(responseMessage);
+		} else {
+			LOG.log(Level.WARNING, "Unkown message type.", message);
 		}
 	}
 
 	protected void handleResponse(ResponseMessage responseMessage) {
-		PendingRequestInfo pendingRequestInfo;
+		PendingRequestInfo requestInfo;
 		synchronized (sentRequestMap) {
-			pendingRequestInfo = sentRequestMap.remove(responseMessage.getId());
+			requestInfo = sentRequestMap.remove(responseMessage.getId());
 		}
-		if (pendingRequestInfo == null) {
+		if (requestInfo == null) {
+			// We have no pending request information that matches the id given in the response
 			LOG.log(Level.WARNING, "Unmatched response message: " + responseMessage);
 		} else {
 			try {
-				pendingRequestInfo.responseHandler.accept(responseMessage);
-			} catch (RuntimeException e) {
+				if (responseMessage.getError() != null)
+					// The remote service has replied with an error
+					requestInfo.future.completeExceptionally(new ResponseErrorException(responseMessage.getError()));
+				else
+					// The remote service has replied with a result object
+					requestInfo.future.complete(responseMessage.getResult());
+			} catch (Exception e) {
 				LOG.log(Level.WARNING, "Handling response threw an exception: " + responseMessage, e);
 			}
 		}
 	}
 
 	protected void handleNotification(NotificationMessage notificationMessage) {
 		if (!handleCancellation(notificationMessage)) {
+			// Forward the notification to the local endpoint
 			try {
 				localEndpoint.notify(notificationMessage.getMethod(), notificationMessage.getParams());
-			} catch (RuntimeException e) {
-				LOG.log(Level.WARNING, "Notification threw an exception: " + notificationMessage, e);
+			} catch (Exception exception) {
+				LOG.log(Level.WARNING, "Notification threw an exception: " + notificationMessage, exception);
```

Comment:
Though I suppose you may want a separate exception handler because notification can’t return an Error. For instance, my exception handler logs the exception. But now I have no way to do this for notifications.

Reply:
The RemoteEndpoint’s exception handler is used only for creating a ResponseError as a reply to a RequestMessage. If you want to customize LSP4J's logging, use the API given by `java.util.logging.Logger`.